ui/src/asm_cleanup.rs

```diff
@@ -107,7 +110,28 @@ pub fn filter_asm(block: &str) -> String {
         }
     }
 
-    let used_labels: HashSet<_> = labels.intersection(&opcode_operands).collect();
+    let mut used_labels: HashSet<_> = labels.intersection(&opcode_operands).collect();
+    
+    // We only include labels ref'd by data dirs if the data dir is part of a used label
+    // Check all labels that include data dirs
+    // If label is used, treat all referenced labels from its data dirs as used
+    // Repeat this to check for any addt'l labels found until no more are located
+    let mut checked_labels = HashSet::new();
+    const MAX_ITERATIONS: u32 = 100; // Safeguard against hangs, should not be an issue but better to be safe
+
+    for _ in 0..MAX_ITERATIONS {
+        let mut new_used_label_found = false;
+
+        for (label, references) in &data_dir_label_refs {
+            if used_labels.contains(label) && checked_labels.insert(label) {
+                new_used_label_found = true;
+                used_labels.extend(references);
+            }
+            if !new_used_label_found {
+                break;
+            }
+        }
+    }
```

Comment:
I'm afraid I've busted your logic here... or at least I don't understand it anymore. 

On the first iteration of the `for` loop, `new_used_label_found` is false. So long as we find the current `label`, the loop will continue, but as soon as we hit the first one, the loop stops. I think it's equivalent to this:

```rust
for _ in 0..MAX_ITERATIONS {
    for (label, references) in &data_dir_label_refs {
        if used_labels.contains(label) && checked_labels.insert(label) {
            new_used_label_found = true;
            used_labels.extend(references);
        } else {
            break;
        }
    }
}
```

This seems a little strange that we'd exit the inner loop like that; maybe the if should be outside the `for`?

---

Is this algorithm essentially trying to find the reachable nodes of a graph starting from a set of nodes? If so, you might want to look at [`petgraph::GraphMap`](https://docs.rs/petgraph/0.4.11/petgraph/graphmap/struct.GraphMap.html) instead of rebuilding a graph by hand. It may even have an algorithm for finding the reachable nodes...

Reply:
Sloppy mistake, my apologies.

Yes, your summary is correct.  I'll look into petgraph, it should be an interesting to try something new.

Reply:
That was much less painful than I was expecting, thanks for the suggestion!